MonoIgnoreThen doesn't take cancellation into account when chaining two callable monos

Hello, I'm not entirely sure if it is an issue or a planned change, but after upgrading dependency from 3.4.4 to reactor 3.4.5 (https://github.com/reactor/reactor-core/releases/tag/v3.4.5) timeout for a chain of monos stopped working as expected.

## Expected Behavior
In the sample below there's a chain of three runnable monos, each taking about 1000mls to be executed.
Process timeout is set to 1500 mls, meaning that timeout will be reached before all runnables are completed. 

Expected behaviour: timeout exception thrown, runnable number 3 is not invoked (because timeout is already reached)

## Actual Behavior
 Timeout exception thrown, runnable number 3 is invoked. 

## Steps to Reproduce

```java

  @Test
  public void test() {
    Set<Integer> runSteps = new HashSet<>();
    Exception exception = assertThrows(Exception.class, () ->
        Mono.fromRunnable(() -> run(1, runSteps))
          .then(Mono.fromRunnable(() -> run(2, runSteps)))
          .then(Mono.fromRunnable(() -> run(3, runSteps)))
          .timeout(Duration.ofMillis(1500))
        .block());
    assertTrue(exception.getCause() instanceof TimeoutException);
    assertEquals(2, runSteps.size());
  }

  private void run(int i, Set<Integer> runSteps) {
    try {
      System.out.println("Running stage " + i);
      Thread.sleep(1000);
      runSteps.add(i);
      System.out.println("Running stage " + i + " done");
    } catch (InterruptedException e) {
      e.printStackTrace();
    }
  }

```


* Reactor version(s) used: 3.4.5 (works fine with 3.4.4)
* Other relevant libraries versions (eg. `netty`, ...): build via springboot dependency springboot-webflux 2.4.5
* JVM version (`java -version`): 1.8
* OS and version (eg `uname -a`): Win10

slightly modifying the snippet, with logs we can see that the runnables / then chain is cancelled and the timeout triggers at the right time (middle of run 2), yet run 3 still starts:

```java
	@Test
	void test() {
		List<Integer> runSteps = new CopyOnWriteArrayList<>();
		assertThatExceptionOfType(RuntimeException.class).isThrownBy(() ->
			Mono.fromRunnable(() -> run(1, runSteps))
				.then(Mono.fromRunnable(() -> run(2, runSteps)))
				.then(Mono.fromRunnable(() -> run(3, runSteps)))
				.log("runnables")
				.timeout(Duration.ofMillis(1500))
				.log("timeout")
				.block())
				.withCauseExactlyInstanceOf(TimeoutException.class);
		assertThat(runSteps).containsExactly(1);
	}

	private void run(int i, Collection<Integer> runSteps) {
		try {
			System.out.println("Running stage " + i + " in thread " + Thread.currentThread().getName());
			Thread.sleep(1000);
			runSteps.add(i);
			System.out.println("Running stage " + i + " done");
		} catch (InterruptedException e) {
			System.err.println("Stage " + i + " has been interrupted / cancelled");
		}
	}
```
logs
```
17:18:53.952 [main] INFO  runnables - onSubscribe(MonoIgnoreThen.ThenIgnoreMain)
17:18:53.957 [main] INFO  timeout - onSubscribe(SerializedSubscriber)
17:18:53.958 [main] INFO  timeout - request(unbounded)
17:18:53.958 [main] INFO  runnables - request(unbounded)
Running stage 1 in thread main
Running stage 1 done
Running stage 2 in thread main
17:18:55.469 [parallel-1] INFO  runnables - cancel()
17:18:55.470 [parallel-1] ERROR timeout - onError(java.util.concurrent.TimeoutException: Did not observe any item or terminal signal within 1500ms in 'log' (and no fallback has been configured))
17:18:55.472 [parallel-1] ERROR timeout - 
java.util.concurrent.TimeoutException: Did not observe any item or terminal signal within 1500ms in 'log' (and no fallback has been configured)
	at reactor.core.publisher.FluxTimeout$TimeoutMainSubscriber.handleTimeout(FluxTimeout.java:295)
	at reactor.core.publisher.FluxTimeout$TimeoutMainSubscriber.doTimeout(FluxTimeout.java:280)
	at reactor.core.publisher.FluxTimeout$TimeoutTimeoutSubscriber.onNext(FluxTimeout.java:419)
	at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onNext(FluxOnErrorResume.java:79)
	at reactor.core.publisher.MonoDelay$MonoDelayRunnable.propagateDelay(MonoDelay.java:271)
	at reactor.core.publisher.MonoDelay$MonoDelayRunnable.run(MonoDelay.java:286)
	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:68)
	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:28)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Running stage 2 done
Running stage 3 in thread main
Running stage 3 done
17:18:56.970 [main] INFO  runnables - onComplete()

java.lang.AssertionError: 
Expecting actual:
  [1, 2, 3]
to contain exactly (and in same order):
  [1]
```

On the other hand, if we `.hide()` the _callable_ nature of what's chained in the `then`, like so:
```java
//...
			Mono.fromRunnable(() -> run(1, runSteps))
				.then(Mono.fromRunnable(() -> run(2, runSteps))
					.hide())
				.then(Mono.fromRunnable(() -> run(3, runSteps))
					.hide())
				.log("runnables")
//...
```
then run 3 doesn't start:
```
17:20:53.826 [main] INFO  runnables - onSubscribe(MonoIgnoreThen.ThenIgnoreMain)
17:20:53.840 [main] INFO  timeout - onSubscribe(SerializedSubscriber)
17:20:53.842 [main] INFO  timeout - request(unbounded)
17:20:53.842 [main] INFO  runnables - request(unbounded)
Running stage 1 in thread main
Running stage 1 done
Running stage 2 in thread main
17:20:55.362 [parallel-1] INFO  runnables - cancel()
17:20:55.362 [parallel-1] ERROR timeout - onError(java.util.concurrent.TimeoutException: Did not observe any item or terminal signal within 1500ms in 'log' (and no fallback has been configured))
17:20:55.365 [parallel-1] ERROR timeout - 
java.util.concurrent.TimeoutException: Did not observe any item or terminal signal within 1500ms in 'log' (and no fallback has been configured)
	at reactor.core.publisher.FluxTimeout$TimeoutMainSubscriber.handleTimeout(FluxTimeout.java:295)
	at reactor.core.publisher.FluxTimeout$TimeoutMainSubscriber.doTimeout(FluxTimeout.java:280)
	at reactor.core.publisher.FluxTimeout$TimeoutTimeoutSubscriber.onNext(FluxTimeout.java:419)
	at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onNext(FluxOnErrorResume.java:79)
	at reactor.core.publisher.MonoDelay$MonoDelayRunnable.propagateDelay(MonoDelay.java:271)
	at reactor.core.publisher.MonoDelay$MonoDelayRunnable.run(MonoDelay.java:286)
	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:68)
	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:28)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Running stage 2 done

java.lang.AssertionError: 
Expecting actual:
  [1, 2]
to contain exactly (and in same order):
  [1, 2]
```

Note that `Thread.sleep` is still undesirable here. If you actually ensure that these sleeps are executed on eg. the `Schedulers.boundedElastic()`, run2 would even get interrupted!

--- a/reactor-core/src/main/java/reactor/core/publisher/MonoIgnoreThen.java
+++ b/reactor-core/src/main/java/reactor/core/publisher/MonoIgnoreThen.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2016-2021 VMware Inc. or its affiliates, All Rights Reserved.
+ * Copyright (c) 2016-2022 VMware Inc. or its affiliates, All Rights Reserved.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -219,6 +219,10 @@ final class MonoIgnoreThen<T> extends Mono<T> implements Scannable {
                 if (i == a.length) {
                     Mono<T> m = this.lastMono;
                     if (m instanceof Callable) {
+                        if (isCancelled(this.state)) {
+                            //NB: in the non-callable case, this is handled by activeSubscription.cancel()
+                            return;
+                        }
                         T v;
                         try {
                             v = ((Callable<T>)m).call();
@@ -240,6 +244,10 @@ final class MonoIgnoreThen<T> extends Mono<T> implements Scannable {
                     final Publisher<?> m = a[i];
 
                     if (m instanceof Callable) {
+                        if (isCancelled(this.state)) {
+                            //NB: in the non-callable case, this is handled by activeSubscription.cancel()
+                            return;
+                        }
                         try {
                             Operators.onDiscard(((Callable<?>) m).call(), currentContext());
                         }

--- a/reactor-core/src/test/java/reactor/core/publisher/MonoIgnoreThenTest.java
+++ b/reactor-core/src/test/java/reactor/core/publisher/MonoIgnoreThenTest.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2016-2021 VMware Inc. or its affiliates, All Rights Reserved.
+ * Copyright (c) 2016-2022 VMware Inc. or its affiliates, All Rights Reserved.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -17,6 +17,12 @@
 package reactor.core.publisher;
 
 import java.time.Duration;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.TimeoutException;
 
 import org.junit.jupiter.api.Nested;
 import org.junit.jupiter.api.Tag;
@@ -30,10 +36,87 @@ import reactor.core.scheduler.Schedulers;
 import reactor.test.StepVerifier;
 import reactor.test.publisher.TestPublisher;
 import reactor.test.subscriber.AssertSubscriber;
+import reactor.util.Logger;
+import reactor.util.Loggers;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
 
 class MonoIgnoreThenTest {
+
+	private static final Logger LOGGER = Loggers.getLogger(MonoIgnoreThenTest.class);
+
+	@Test
+	void multipleThenChainedCancelInTheMiddle() throws InterruptedException {
+		List<Integer> startedSteps = new CopyOnWriteArrayList<>();
+		List<Integer> completedSteps = new CopyOnWriteArrayList<>();
+		List<Integer> cancelledSteps = new CopyOnWriteArrayList<>();
+
+		Mono<?> run1 = Mono.fromRunnable(() -> multipleThenStep(1, startedSteps, completedSteps, cancelledSteps)).subscribeOn(Schedulers.boundedElastic());
+		Mono<?> run2 = Mono.fromRunnable(() -> multipleThenStep(2, startedSteps, completedSteps, cancelledSteps)).subscribeOn(Schedulers.boundedElastic());
+		Mono<?> run3 = Mono.fromRunnable(() -> multipleThenStep(3, startedSteps, completedSteps, cancelledSteps)).subscribeOn(Schedulers.boundedElastic());
+
+		assertThat(Arrays.asList(run1, run2, run3)).as("smoke test: no callables").noneMatch(m -> m instanceof Callable);
+
+		assertThatExceptionOfType(RuntimeException.class).isThrownBy(() ->
+				run1.then(run2)
+					.then(run3)
+					.timeout(Duration.ofMillis(750))
+					.block())
+			.withCauseExactlyInstanceOf(TimeoutException.class);
+
+		//we give 1.5s total in case run3 incorrectly starts or run2 incorrectly completes
+		Thread.sleep(750);
+		//we verify that run3 didn't even start
+		assertThat(startedSteps).as("started steps").containsExactly(1, 2);
+		//in this configuration the sleep is interrupted so run2 "done" should not happen
+		assertThat(completedSteps).as("done steps").containsExactly(1);
+		assertThat(cancelledSteps).as("interruptions").containsExactly(2);
+	}
+
+	@Test
+	void multipleThenChainedCancelInTheMiddle_withCallables() throws InterruptedException {
+		List<Integer> startedSteps = new CopyOnWriteArrayList<>();
+		List<Integer> completedSteps = new CopyOnWriteArrayList<>();
+		List<Integer> cancelledSteps = new CopyOnWriteArrayList<>();
+
+		Mono<?> run1 = Mono.fromRunnable(() -> multipleThenStep(1, startedSteps, completedSteps, cancelledSteps));
+		Mono<?> run2 = Mono.fromRunnable(() -> multipleThenStep(2, startedSteps, completedSteps, cancelledSteps));
+		Mono<?> run3 = Mono.fromRunnable(() -> multipleThenStep(3, startedSteps, completedSteps, cancelledSteps));
+
+		assertThat(Arrays.asList(run1, run2, run3)).as("smoke test: all callables").allMatch(m -> m instanceof Callable);
+
+		assertThatExceptionOfType(RuntimeException.class).isThrownBy(() ->
+				run1.then(run2)
+					.then(run3)
+				.timeout(Duration.ofMillis(750))
+				.block())
+				.withCauseExactlyInstanceOf(TimeoutException.class);
+
+		//we give 1.5s total in case run3 incorrectly starts
+		Thread.sleep(750);
+		//we verify that run3 didn't even start
+		assertThat(startedSteps).as("started steps").containsExactly(1, 2);
+		//unfortunately in this configuration the sleep cannot really be interrupted so the "done" will still happen
+		assertThat(completedSteps).as("done steps").containsExactly(1, 2);
+		assertThat(cancelledSteps).as("no interruptions").isEmpty();
+	}
+
+	private void multipleThenStep(int i, Collection<Integer> startedSteps, Collection<Integer> completedSteps, Collection<Integer> cancelledSteps) {
+		startedSteps.add(i);
+		try {
+			LOGGER.debug("Running stage {}", i);
+			//this sleep is problematic usually, but done here on purpose in order to have a Callable Mono
+			// that induces a sufficient delay for timeout to trigger
+			Thread.sleep(500);
+			LOGGER.debug("Running stage {} done", i);
+			completedSteps.add(i);
+		} catch (InterruptedException e) {
+			cancelledSteps.add(i);
+			LOGGER.debug("Stage {} has been interrupted / cancelled", i);
+		}
+	}
+
 
 	@Nested
 	class ThenIgnoreVariant {

